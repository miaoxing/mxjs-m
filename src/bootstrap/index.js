import config from '@/config';
import {setApiUrl, setOnBeforeHttp} from '@mxjs/taro';
import Taro from '@tarojs/taro';
import $ from 'miaoxing';
import '@/components/Page';

// bootstrap: 引入自定义的样式主题
import 'fower-preset-miaoxing';

const login = async () => {
  // 1. 调用登录获取 code
  const res = await Taro.login();
  if (!res.code) {
    $.alert('微信登录失败：' + res.errMsg);
    return;
  }

  // 2. 去后台换为 OpenID，返回登录 token
  const {ret} = await $.http({
    url: $.apiUrl('wechat-mp/login'),
    method: 'post',
    // 跳过以免循环
    skipBeforeEvent: true,
    data: {
      code: res.code,
    },
  });
  if (ret.isErr()) {
    $.alert(ret.message);
    return;
  }

  // 3. 记录 token
  Taro.setStorageSync('token', ret.token);
};

const bootstrap = async () => {
  // 设置请求接口的地址
  setApiUrl(config.apiUrl);

  // 请求登录
  const loginResult = login();
  setOnBeforeHttp(async () => {
    await loginResult;
  });
};

export default bootstrap;
