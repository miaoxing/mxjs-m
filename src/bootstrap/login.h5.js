import Taro from '@tarojs/taro';
import $ from 'miaoxing';
import qs from 'query-string';
import appendUrl from 'append-url';

/**
 * 微信服务号登录
 *
 * 暂未支持非微信环境
 */
export default async () => {
  if (!navigator.userAgent.includes('MicroMessenger')) {
    return;
  }

  if (Taro.getStorageSync('token')) {
    return;
  }

  const code = $.req('code');
  const state = $.req('state');

  // 授权回调
  if (code && typeof state !== 'undefined') {
    const {ret} = await $.post({
      url: $.apiUrl('wechat-oa/login'),
      data: {code, state, url: window.location.href},
    });
    if (ret.isSuc()) {
      Taro.setStorageSync('token', ret.token);
      const {code: codeIgnore, state: stateIgnore, ...params} = qs.parse(window.location.search);
      // 跳转移除当前地址中的 code 和 state 参数
      // noinspection ES6MissingAwait 实际不会 resolve
      Taro.redirectTo({
        url: appendUrl(window.location.pathname, params),
      });
      return;
    }

    if (ret.retryUrl) {
      window.location = ret.retryUrl;
    } else {
      $.alert(ret.message);
    }
    return;
  }

  // 跳转到授权地址
  const {ret} = await $.get($.apiUrl('wechat-oa/login', {url: window.location.href}));
  if (ret.isErr()) {
    // 忽略错误，例如未设置服务号
    return;
  }
  window.location = ret.url;
};
