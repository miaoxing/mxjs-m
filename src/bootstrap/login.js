module.exports = require('./login.' + (process.env.TARO_ENV || 'weapp'));
