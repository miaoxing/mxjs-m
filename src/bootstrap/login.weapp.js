import Taro from '@tarojs/taro';
import $ from 'miaoxing';

export default async () => {
  // 1. 调用登录获取 code
  const res = await Taro.login();
  if (!res.code) {
    $.alert('微信登录失败：' + res.errMsg);
    return;
  }

  // 2. 去后台换为 OpenID，返回登录 token
  const {ret} = await $.http({
    url: $.apiUrl('wechat-mp/login'),
    method: 'post',
    // 跳过以免循环
    skipBeforeEvent: true,
    data: {
      code: res.code,
    },
  });
  if (ret.isErr()) {
    $.alert(ret.message);
    return;
  }

  // 3. 记录 token
  Taro.setStorageSync('token', ret.token);
};
