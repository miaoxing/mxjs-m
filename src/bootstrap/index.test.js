import bootstrap from './index';
import Taro from '@tarojs/taro';
import $, {Ret} from 'miaoxing';
import {createPromise} from '@mxjs/test';

describe('bootstrap', () => {
  test('login by mp', async () => {
    Taro.login = jest.fn().mockResolvedValueOnce({
      code: 'test-code',
    });

    $.http = jest.fn().mockResolvedValueOnce({
      ret: Ret.suc({token: 'test-token'}),
    });
    
    const promise = createPromise();
    Taro.setStorageSync = jest.fn().mockImplementationOnce(() => promise.resolve());

    await bootstrap();
    await promise;

    expect(Taro.login.mock.calls).toMatchSnapshot();
    expect($.http.mock.calls).toMatchSnapshot();
    expect(Taro.setStorageSync.mock.calls).toMatchSnapshot();
  });

  test('login by mp: Taro login fail', async () => {
    const promise = createPromise();
    Taro.login = jest.fn().mockImplementationOnce(() => promise.resolve({
      errMsg: '登录失败',
    }));

    $.alert = jest.fn();

    await bootstrap();
    await promise;

    expect(Taro.login.mock.calls).toMatchSnapshot();
    expect($.alert.mock.calls).toMatchSnapshot();
  });

  test('login by mp: HTTP fail', async () => {
    Taro.login = jest.fn().mockResolvedValueOnce({
      code: 'test-code',
    });

    const promise = createPromise();
    $.http = jest.fn().mockImplementationOnce(() => promise.resolve({
      ret: Ret.err('HTTP error'),
    }));

    $.alert = jest.fn();

    await bootstrap();
    await promise;

    expect(Taro.login.mock.calls).toMatchSnapshot();
    expect($.http.mock.calls).toMatchSnapshot();
    expect($.alert.mock.calls).toMatchSnapshot();
  });
});
