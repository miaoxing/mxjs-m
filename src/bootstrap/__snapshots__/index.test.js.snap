// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bootstrap login by mp 1`] = `
Array [
  Array [],
]
`;

exports[`bootstrap login by mp 2`] = `
Array [
  Array [
    Object {
      "data": Object {
        "code": "test-code",
      },
      "method": "post",
      "skipBeforeEvent": true,
      "url": "/m-api/wechat-mp/login",
    },
  ],
]
`;

exports[`bootstrap login by mp 3`] = `
Array [
  Array [
    "token",
    "test-token",
  ],
]
`;

exports[`bootstrap login by mp: HTTP fail 1`] = `
Array [
  Array [],
]
`;

exports[`bootstrap login by mp: HTTP fail 2`] = `
Array [
  Array [
    Object {
      "data": Object {
        "code": "test-code",
      },
      "method": "post",
      "skipBeforeEvent": true,
      "url": "/m-api/wechat-mp/login",
    },
  ],
]
`;

exports[`bootstrap login by mp: HTTP fail 3`] = `Array []`;

exports[`bootstrap login by mp: Taro login fail 1`] = `
Array [
  Array [],
]
`;

exports[`bootstrap login by mp: Taro login fail 2`] = `
Array [
  Array [
    "微信登录失败：登录失败",
  ],
]
`;
